### ISSUE TITLE
Add ability to manually pretranslate strings

### ISSUE THREAD

**Issue by mathjazz** [mozilla/pontoon#2790]
Teams might want to benefit from pretranslation capabilities on projects for which pretranslation is not enabled (by the admins).

Manual pretranslation would be triggered by Translators, either for the entire project or for a subset of strings. We could also add it as an additional batch action.

We should take into account that pretranslation is a background process that can take some time and modify the UI accordingly while it's in progress.

**Comment by mathjazz** [comment]
We can add this feature as a batch action for locales that have pretranslation enabled.

While the pretranslation is in progress, we should make the editor disabled for the selected strings and add a notice, similar to the logged-out or read-only mode.

**Comment by MundiaNderi** [comment]
I had a look at the existing pretranslation documentation, code, and what was discussed in the meeting. My thought process:

- Check whether a locale is supported by translation engines
     - If supported, enable pretranslation under the following conditions:
             - String has landed and has no translation yet
             - Skip strings that already have any translation: approved, suggestion, or previously pretranslated. Open to discussion if there's a reason to pretranslate over existing suggestions instead.
             - Skip if the locale is marked read-only for the project

-  Pretranslation should run as a background task.  While it's running:
      - Disable the editor for the selected strings and show a notice

Questions:

- Should we cap the number of strings that can be pretranslated at once, given https://github.com/mozilla/pontoon/blob/16672a885ff1cc6cc706558271d757bbcd89d183/pontoon/pretranslation/pretranslate.py#L278
- For locales not supported by the translation engines, should we show a "not supported for this locale" message rather than a silent no-op?

**Comment by mathjazz** [comment]
> I had a look at the existing pretranslation documentation, code, and what was discussed in the meeting. My thought process:
>         - Skip strings that already have any translation: approved, suggestion, or previously pretranslated. Open to discussion if there's a reason to pretranslate over existing suggestions instead.

Agreed, which matches the the existing pretranslation behaviour. Note that we should also pretranslate strings that have existing translations, but they are all rejected and submitted by humans.

@flodolo Thoughts?

> Should we cap the number of strings that can be pretranslated at once, given https://github.com/mozilla/pontoon/blob/16672a885ff1cc6cc706558271d757bbcd89d183/pontoon/pretranslation/pretranslate.py#L278

Not sure what you had in mind with the code you referenced?

> For locales not supported by the translation engines, should we show a "not supported for this locale" message rather than a silent no-op?

I wouldn't even show the pretranslation section in the batch actions panel for those locales.

**Comment by flodolo** [comment]
> > I had a look at the existing pretranslation documentation, code, and what was discussed in the meeting. My thought process:
> > - Skip strings that already have any translation: approved, suggestion, or previously pretranslated. Open to discussion if there's a reason to pretranslate over existing suggestions instead.
> 
> Agreed, which matches the the existing pretranslation behaviour. Note that we should also pretranslate strings that have existing translations, but they are all rejected and submitted by humans.
> 
> [@flodolo](https://github.com/flodolo) Thoughts?

I would mimic what pretranslation is currently doing, and keep a change of behavior across the board for a separate issue/discussion.

IMO, we should pretranslate even if there are pending suggestions, as they serve a different purpose. If we're pretranslating, it's because we want that content to end up in VCS, not just stay in Pontoon. The hard question is what happens if the pretranslation matches an existing suggestion.

**Comment by mathjazz** [comment]
> IMO, we should pretranslate even if there are pending suggestions, as they serve a different purpose. If we're pretranslating, it's because we want that content to end up in VCS, not just stay in Pontoon. The hard question is what happens if the pretranslation matches an existing suggestion.

Yeah, let's track this separately.

**Comment by flodolo** [comment]
This batch action should not be available to locale managers/translators.

> Check whether a locale is supported by translation engines

We enable pretranslation on a locale **AND** project basis. 
https://mozilla-l10n.github.io/localizer-documentation/tools/pretranslation.html#opt-in-guidelines

Providing this feature to managers and translators defies the point of having such guidelines. We're not at a point where I feel confident in enabling pretranslation across the board for a locale, maybe once we have better linting in Pontoon, and we have experimented with alternative back-ends.

Having this batch action is not particularly useful on a project where pretranslation is already enabled, since it's supposed to pretranslate all resources.  It only makes sense if we change the behavior, i.e. pretranslate where we already have suggestions, and we decided above that's a different conversation.

This action would make sense for admins — or even better, the manually curated `pretranslators` group — to pretranslate strings in a project as a one-off. The same result can be achieved in `/admin`, hence the suggestion to make it available to the group:
1. Enable pretranslation for that locale
2. Pretranslate with the button
3. Disable pretranslation for that locale

**Comment by mathjazz** [comment]
> We enable pretranslation on a locale **AND** project basis.
> https://mozilla-l10n.github.io/localizer-documentation/tools/pretranslation.html#opt-in-guidelines

Phew, I missed the project-level criteria.

Regarding the use case: I often wish I could use the pretranslation feature on parts of larger project where pretranslation **is not enabled** as a whole, e.g. on low-priority Firefox strings or long Mozilla.org strings.
